lane.py

```diff
@@ -387,916 +375,852 @@ def has_next_page(self):
+        apply_custom_filters() hook method.
+        """
+        # In general, we only show books that are ready to be delivered
+        # to patrons.
+        qu = self.only_show_ready_deliverable_works(_db, qu, work_model)
+
+        # This method applies whatever filters are necessary to implement
+        # the rules of this particular WorkList.
+        qu, distinct = self.apply_bibliographic_filters(
+            _db, qu, work_model, featured
+        )
+        if not qu:
+            # apply_bibliographic_filters() may return a null query to
+            # indicate that the WorkList should not exist at all.
+            return None
 
-        If you set audience to Young Adult but age_range to 16-18,
-        you're saying that books for 18-year-olds (i.e. adults) are
-        okay.
+        if facets:
+            qu = facets.apply(_db, qu, work_model, distinct=distinct)
+        elif distinct:
+            # Something about the query makes it possible that the same
+            # book might show up twice. We set the query as DISTINCT
+            # to avoid this possibility.
+            qu = qu.distinct()
 
-        If you set age_range to Young Adult but age_range to 12-15, you're
-        saying that books for 12-year-olds (i.e. children) are
-        okay.
+        if pagination:
+            qu = pagination.apply(qu)
+        return qu
+
+    def apply_bibliographic_filters(self, _db, qu, work_model, featured=False):
+        """Filter out books whose bibliographic metadata doesn't match
+        what we're looking for.
         """
-        if not audiences:
-            if self.parent:
-                audiences = self.parent.audiences
-            else:
-                audiences = []
-        if isinstance(audiences, basestring):
-            audiences = [audiences]
-        if isinstance(audiences, set):
-            audiences = audiences
-        else:
-            audiences = set(audiences)
-        if not age_range:
-            return audiences
+        # Audience and language restrictions are common to all
+        # WorkLists. (So are genre and collection restrictions, but those
+        # were applied back in works().)
+        qu = self.apply_audience_filter(_db, qu, work_model)
+        if self.languages:
+            qu = qu.filter(work_model.language.in_(self.languages))
+        if self.genre_ids:
+            qu = qu.filter(work_model.genre_id.in_(self.genre_ids))
+        return qu, False
 
-        if not isinstance(age_range, list):
-            age_range = [age_range]
+    def apply_audience_filter(self, _db, qu, work_model):
+        """Make sure that only Works classified under this lane's
+        allowed audiences are returned.
+        """
+        if not self.audiences:
+            return qu
+        qu = qu.filter(work_model.audience.in_(self.audiences))
+        if (Classifier.AUDIENCE_CHILDREN in self.audiences
+            or Classifier.AUDIENCE_YOUNG_ADULT in self.audiences):
```

Comment:
@leonardr Shouldn't this hack be applied if adult is not in the audiences, rather than if children or ya is?

If the WorkList is only for children or ya, you want to filter out the Gutenberg books, but if it's for any audience, they can stay.

Reply:
It's pretty theoretical, but I'm concerned about a scenario like the Open Ebooks "High School" lane, which has audience=(Young Adult, Adult) but which still shouldn't have these books.